src/Grenade/Core/Network.hs

```diff
@@ -179,19 +240,120 @@ instance (SingI i, SingI o, Layer x i o, Serialize x, Serialize (Network xs (o '
   put (x :~> r) = put x >> put r
   get = (:~>) <$> get <*> get
 
-
 -- | Ultimate composition.
 --
 --   This allows a complete network to be treated as a layer in a larger network.
-instance CreatableNetwork sublayers subshapes => UpdateLayer (Network sublayers subshapes) where
+instance UpdateLayer (Network sublayers subshapes) where
   type Gradient (Network sublayers subshapes) = Gradients sublayers
-  runUpdate    = applyUpdate
-  createRandom = randomNetwork
+  runUpdate = applyUpdate
+  runSettingsUpdate = applySettingsUpdate
+
+instance FoldableGradient (Gradients '[]) where
+  mapGradient _ GNil = GNil
+  squaredSums GNil = []
+
+instance (FoldableGradient (Gradient x), FoldableGradient (Gradients xs)) => FoldableGradient (Gradients (x ': xs)) where
+  mapGradient f (x :/> xs) =
+    let x' = mapGradient f x
+        xs' = mapGradient f xs
+     in x' :/> xs'
+  squaredSums (x :/> xs) = squaredSums x ++ squaredSums xs
+
+-- | Get the L2 Norm of a Foldable Gradient.
+l2Norm :: (FoldableGradient x) => x -> RealNum
+l2Norm grad = sqrt (sum $ squaredSums grad)
+
+-- | Clip the gradients by the global norm.
+clipByGlobalNorm :: (FoldableGradient (Gradients xs)) => Double -> Gradients xs -> Gradients xs
+clipByGlobalNorm c grads =
+  let divisor = sqrt $ sum $ squaredSums grads
+   in if divisor > c
+        then mapGradient (* (c / divisor)) grads
+        else grads
+
+
+instance CreatableNetwork sublayers subshapes => RandomLayer (Network sublayers subshapes) where
+  createRandomWith = randomNetworkWith
+
 
 -- | Ultimate composition.
 --
 --   This allows a complete network to be treated as a layer in a larger network.
-instance (CreatableNetwork sublayers subshapes, i ~ (Head subshapes), o ~ (Last subshapes)) => Layer (Network sublayers subshapes) i o where
+instance (i ~ (Head subshapes), o ~ (Last subshapes)) => Layer (Network sublayers subshapes) i o where
   type Tape (Network sublayers subshapes) i o = Tapes sublayers subshapes
   runForwards  = runNetwork
   runBackwards = runGradient
+
+
+--------------------------------------------------
+
+-- | Grenade Num class.
+--
+-- This allows for instance scalar multiplication of the weights, which is useful for slowly adapting networks, e.g. NN'
+-- <- \tau * NN' + (1-\tau) * NN. Or one could sum up some gradients in parallel and apply them at once: @applyUpdate lp
+-- net $ foldl1 (|+) ...@aq.
+class GNum a where
```

Comment:
This are the operations of a semiring. If the laws hold, it might be worth just using the `semirings` package.

Reply:
True, although I am wondering if we not rather should drop the `gFromRational` function. I am not sure if it use more useful or dangerous.

Reply:
If it's not necessary I would try not to include it. I'm absolutely for using known mathematical structures over ad-hoc ones.